# Linkage disequilibrium

In this module, we'll use DNA sequencing data from human populations to assess linkage between two genetic variants.

#### Learning objectives

After completing this chapter, you'll be able to:

1. Define linkage disequilibrium.
2. Manually calculate $D$, $D'$, and $r^2$ from genotype data.
3. Understand the differences in interpretation for different LD statistics.
4. Explain how LD both benefits and limits genetics studies.


***


## Setup

### Data

Load the pre-processed VCF data by running the code chunk below.

```{r}
# read data
haplotypes <- read.table("snp_haplotypes.txt",
                         header = TRUE)

# preview data
head(haplotypes)
```


***


## Are these SNPs in LD?

Run `table` on each SNP column to see which alleles exist at SNP1 and SNP2.

```{r}
table(haplotypes$snp1_allele)
table(haplotypes$snp2_allele)
```

Use the `table` function to count the occurrences of the four possible haplotypes.

```{r}
table(haplotypes$snp1_allele, haplotypes$snp2_allele)
```

Perform a Fisher's exact test to test for evidence of LD:

```{r}
fisher.test(table(haplotypes$snp1_allele, haplotypes$snp2_allele))
```


***


## Measuring LD with D

$$
D = h_{12} - p_1*p_2
$$

Plug in the corresponding probabilities from the table to calculate `D`:

```{r}
# define our probabilities of interest
h <- ____
p1 <- ____
p2 <- ____

# calculate D
D <- h - (p1 * p2)
D
```


***


## Measuring LD with D'

$$
D' = \frac{D}{\mathrm{max}(-p_1 p_2, -(1-p_1)(1-p_2))}, \mathrm{\:for\:} D < 0 \\
D' = \frac{D}{\mathrm{min}(p_1 (1-p_2), p_2(1-p_1) )}, \mathrm{\:for\:} D > 0
$$

Determine the minimum of $p_1 (1-p_2)$ and $p_2 (1-p_1)$:

```{r}
p1 * (1-p2)
p2 * (1-p1)
```

Use the formula above to calculate D':

```{r}
Dprime <- ____ / ____
Dprime
```



***


## Measuring LD with r^2

$$
r^2 = \frac{D^2}{p_1 (1-p_1) p_2 (1-p_2)}
$$

Use the formula above to calculate r^2:

```{r}
r2 <- ____ / ____
r2
```


***


## LDlink

https://ldlink.nci.nih.gov/


***


## Homework

#### Learning Objectives

* Calculate and interpret LD statistics

#### Assignment

We've subset the VCF from class to show haplotypes for two different pair of SNPs:

* `chr21:15005329` and `chr21:15007704`
* `chr21:15336586` and `chr21:15336794`

```{r}
# read data for first set of SNPs
hw1 <- read.table("snp_haplotypes_hw1.txt", header = TRUE)

# read data for second set of SNPs
hw2 <- read.table("snp_haplotypes_hw2.txt", header = TRUE)
```

**Assignment:** Using the code from class, calculate $D$, $D'$, and $r^2$ for these sets of SNPs. Which alleles are segregating together? What does each LD statistic indicate? (Feel free to check your work on LDpair.)

### First set of SNPs

```{r}
# D
```

```{r}
# D'
```

```{r}
# r^2
```

### Second set of SNPs

```{r}
# D
```

```{r}
# D'
```

```{r}
# r^2
```