<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>2 Course Syllabus | Population Genetics Modeling</title>
  <meta name="description" content="Description of Course" />
  <meta name="generator" content="bookdown 0.39.1 and GitBook 2.6.7" />

  <meta property="og:title" content="2 Course Syllabus | Population Genetics Modeling" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Description of Course" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="2 Course Syllabus | Population Genetics Modeling" />
  
  <meta name="twitter:description" content="Description of Course" />
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="assets/pig.jpg" type="image/x-icon" />
<link rel="prev" href="index.html"/>
<link rel="next" href="lecture-notes.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="assets/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="https://andrew-bortvin.github.io/popGenModeling/"><img src="assets/pig.jpg" style=" width: 80%; padding-left: 40px; padding-top: 8px; vertical-align: top"</a>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Homepage</a></li>
<li class="chapter" data-level="2" data-path="course-syllabus.html"><a href="course-syllabus.html"><i class="fa fa-check"></i><b>2</b> Course Syllabus</a>
<ul>
<li class="chapter" data-level="2.0.1" data-path="course-syllabus.html"><a href="course-syllabus.html#course-description"><i class="fa fa-check"></i><b>2.0.1</b> Course Description</a></li>
<li class="chapter" data-level="2.0.2" data-path="course-syllabus.html"><a href="course-syllabus.html#learning-goals"><i class="fa fa-check"></i><b>2.0.2</b> Learning Goals</a></li>
<li class="chapter" data-level="2.0.3" data-path="course-syllabus.html"><a href="course-syllabus.html#grading"><i class="fa fa-check"></i><b>2.0.3</b> Grading</a></li>
<li class="chapter" data-level="2.0.4" data-path="course-syllabus.html"><a href="course-syllabus.html#tentative-schedule"><i class="fa fa-check"></i><b>2.0.4</b> Tentative Schedule</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="lecture-notes.html"><a href="lecture-notes.html"><i class="fa fa-check"></i><b>3</b> Lecture Notes</a>
<ul>
<li class="chapter" data-level="3.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html"><i class="fa fa-check"></i><b>3.1</b> Introduction to R</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#data-import"><i class="fa fa-check"></i><b>3.1.1</b> Data import</a></li>
<li class="chapter" data-level="3.1.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#the-data-frame"><i class="fa fa-check"></i><b>3.1.2</b> The Data Frame</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="lecture-notes-1.html"><a href="lecture-notes-1.html"><i class="fa fa-check"></i><b>3.2</b> Lecture Notes</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="lecture-notes-1.html"><a href="lecture-notes-1.html#data-import-1"><i class="fa fa-check"></i><b>3.2.1</b> Data import</a></li>
<li class="chapter" data-level="3.2.2" data-path="lecture-notes-1.html"><a href="lecture-notes-1.html#dataframes"><i class="fa fa-check"></i><b>3.2.2</b> DataFrames</a></li>
<li class="chapter" data-level="3.2.3" data-path="lecture-notes-1.html"><a href="lecture-notes-1.html#data-manipulation"><i class="fa fa-check"></i><b>3.2.3</b> Data Manipulation</a></li>
<li class="chapter" data-level="3.2.4" data-path="lecture-notes-1.html"><a href="lecture-notes-1.html#subsetting"><i class="fa fa-check"></i><b>3.2.4</b> Subsetting</a></li>
<li class="chapter" data-level="3.2.5" data-path="lecture-notes-1.html"><a href="lecture-notes-1.html#logical-subsetting"><i class="fa fa-check"></i><b>3.2.5</b> Logical subsetting</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="r-guidebook.html"><a href="r-guidebook.html"><i class="fa fa-check"></i><b>4</b> R Guidebook</a>
<ul>
<li class="chapter" data-level="4.1" data-path="data-types.html"><a href="data-types.html"><i class="fa fa-check"></i><b>4.1</b> Data Types</a></li>
<li class="chapter" data-level="4.2" data-path="variables.html"><a href="variables.html"><i class="fa fa-check"></i><b>4.2</b> Variables</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="variables.html"><a href="variables.html#overwriting-variables"><i class="fa fa-check"></i><b>4.2.1</b> Overwriting Variables</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="math.html"><a href="math.html"><i class="fa fa-check"></i><b>4.3</b> Math</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="math.html"><a href="math.html#order-of-operations"><i class="fa fa-check"></i><b>4.3.1</b> Order of Operations</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="vectors.html"><a href="vectors.html"><i class="fa fa-check"></i><b>4.4</b> Vectors</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="vectors.html"><a href="vectors.html#indexing"><i class="fa fa-check"></i><b>4.4.1</b> Indexing</a></li>
<li class="chapter" data-level="4.4.2" data-path="vectors.html"><a href="vectors.html#modifying-vecotrs"><i class="fa fa-check"></i><b>4.4.2</b> Modifying Vecotrs</a></li>
<li class="chapter" data-level="4.4.3" data-path="vectors.html"><a href="vectors.html#adding-to-vectors"><i class="fa fa-check"></i><b>4.4.3</b> Adding to Vectors</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="data-frames.html"><a href="data-frames.html"><i class="fa fa-check"></i><b>4.5</b> Data Frames</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="data-frames.html"><a href="data-frames.html#subsetting-data-frames"><i class="fa fa-check"></i><b>4.5.1</b> Subsetting Data Frames</a></li>
<li class="chapter" data-level="4.5.2" data-path="data-frames.html"><a href="data-frames.html#logical-indexing"><i class="fa fa-check"></i><b>4.5.2</b> Logical Indexing</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="lecture-slides.html"><a href="lecture-slides.html"><i class="fa fa-check"></i><b>5</b> Lecture Slides</a></li>
<li class="chapter" data-level="" data-path="authors.html"><a href="authors.html"><i class="fa fa-check"></i>Authors</a></li>
<li class="divider"></li>
<p style="text-align:center; font-size: 11pt"> <a href="https://github.com/jhudsl/OTTR_Template" target="blank" > Published with</a> <a href="https://bookdown.org/" target="blank"> <b>bookdown</b></a> (by <a href="http://jhudatascience.org/"> The Johns Hopkins Data Science Lab</a>) </p>
<p style="text-align:center; font-size: 12px;"> <a href="https://github.com/rstudio4edu/rstudio4edu-book/"> Style adapted from: rstudio4edu-book </a> <a href ="https://creativecommons.org/licenses/by/2.0/"> (CC-BY 2.0) </a></p>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Population Genetics Modeling</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<head>
  <meta name="viewport" content="width=device-width,minimum-scale=1.0,maximum-scale=1.0,initial-scale=1.0">
  <link href="assets/style.css" rel="stylesheet">
</head>



<div class="hero-image-container">
  <img class="hero-image" src="assets/ilum.jpg">
</div>
<div id="course-syllabus" class="section level1 hasAnchor" number="2">
<h1><span class="header-section-number">2</span> Course Syllabus<a href="course-syllabus.html#course-syllabus" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p><strong>COMPUTATIONAL APPROACHES TO POPULATION GENETICS</strong><br />
Instructor: Andrew Bortvin<br />
Times: Tuesday, Thursday 3:00PM - 4:30PM<br />
Office Hours: To be determined by when2meet at the start of the semester, or by appointment<br />
Course website: <a href="https://andrew-bortvin.github.io/popGenModeling/" class="uri">https://andrew-bortvin.github.io/popGenModeling/</a></p>
<div id="course-description" class="section level3 hasAnchor" number="2.0.1">
<h3><span class="header-section-number">2.0.1</span> Course Description<a href="course-syllabus.html#course-description" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The size, composition, and genetics of populations fluctuate over time. These fluctuations are the product of dynamics between individuals, the interactions between populations, and the context of a population within a broader ecological landscape. The quantitative tools developed to study population genetics allow biologists to discover the simple fundamental principles that govern these complex systems. This course will introduce the basic theory of population genetics while teaching students the fundamental skills of programming in the R programming language, which will allow them to directly implement and visualize theoretical concepts. Students will model and simulate theoretical populations and analyze population-scale genomic data. This course will examine evolution on a variety of scales, ranging from the competition between cells within a single organism, to population dynamics in conservation biology that span decades, to the evolution of contemporary human populations over hundreds of thousands of years.</p>
</div>
<div id="learning-goals" class="section level3 hasAnchor" number="2.0.2">
<h3><span class="header-section-number">2.0.2</span> Learning Goals<a href="course-syllabus.html#learning-goals" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>By the end of this course, students will be able to:</p>
<p>• outline, using biology theory and terms, how populations grow and interact with each other<br />
• describe the external factors that can cause changes in genetic diversity and use this information to predict how specific demographic scenarios would impact a population<br />
• manipulate, analyze, and visualize data using the R programming language<br />
• describe and interpret common formats used to store genomic data, and implement standard analytic protocols used to analyze these data types</p>
</div>
<div id="grading" class="section level3 hasAnchor" number="2.0.3">
<h3><span class="header-section-number">2.0.3</span> Grading<a href="course-syllabus.html#grading" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>30% Participation<br />
40% Weekly Assignments<br />
30% Final Project</p>
<p><strong>Weekly Assignments</strong></p>
<p>Class sessions will consist of a lecture in which theory is introduced, instructor-led live-coding sessions that implement the models studied in class, and time for students to work independently on assignments that further develop the models designed in class. Most weeks, there will be a take-home assignment that extends concepts studied in class. Students will have a week to complete problem sets. After submission, students will receive feedback from the instructor on each assignment and will have until the end of the semester to submit any revisions necessary.</p>
<p>Each assignment will consist of a set of required exercises that can be completed by students of any coding background. These will be followed by optional, more computationally-focused exercises, which will allow students to examine more intricate evolutionary scenarios and implement more complicated computational models.</p>
<p>Work will be graded on reasonable completion–that is to say, code that demonstrates an understanding of an algorithm and its general implementation will receive full marks, regardless of whether output is exactly correct. Students will also be assessed on the clarity and interpretability of the data visualizations that their code outputs and the accuracy of their responses to short questions prompting biological interpretation of their results.</p>
<p><strong>Google and AI</strong></p>
<p>Googling is always an acceptable way to find answers or help, and I encourage you to utilize it extensively. If you adopt a solution following a Google search, make sure you understand what you incorporate, rather than just copy/paste without comprehension of the logic or code. Google is also a good way to learn more about any error messages you encounter in your code.</p>
<p>You may be familiar with ChatGPT and other large language models. After trying each problem/assignment/task on your own, if you’re still running into issues, feel free to use ChatGPT as you would any other online resource (Google, stack overflow, etc.). Learning how to succinctly describe exactly what you want to accomplish is a skillset in itself, so this can be good practice. If you find code that seems to work (e.g., from Google) but you’re not sure how exactly it works, you can also type it into ChatGPT and ask it to explain what’s happening. As always, please do not submit any code if you are not familiar entirely with how it works; flag it and ask an instructor for assistance. Be aware that ChatGPT might confidently offer an answer that is not correct; so always check the output on your own.</p>
</div>
<div id="tentative-schedule" class="section level3 hasAnchor" number="2.0.4">
<h3><span class="header-section-number">2.0.4</span> Tentative Schedule<a href="course-syllabus.html#tentative-schedule" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>First Week - Introduction to Population Genetics Modeling</strong></p>
<p>August 27: Welcome; Course Overview</p>
<p>August 29: Introduction to R Programming - Working with Data, Plotting</p>
<p><strong>Unit 1: Population Biology</strong><br />
How do population sizes change? Models for one and two populations. Cooperation, Competition, and predation.</p>
<p><strong>Week Two: One Population Models</strong><br />
September 3: The Exponential and Logistic Growth Models</p>
<p>September 5: Density-Dependent Growth</p>
<p><strong>Week Three: Multiple Populations</strong><br />
September 10: Lotka-Volterra dynamics 1: Competition and Cooperation</p>
<p>September 12: Lotka-Volterra dynamics 2: Predation and Parasitism</p>
<p><strong>Week Four: Advanced Topics in Population Biology</strong><br />
September 17: Spatial Models</p>
<p>September 19: Social Evolution and Game Theory</p>
<p><strong>Unit Two: Population Genetics</strong><br />
How do we measure the genetic relationships between individuals? Between Populations? Between Species?<br />
What determines the fate of a genetic element in a population?  
How does the size and demographic history of a population impact its genetic composition?<br />
Biological Simulation</p>
<p><strong>Week Five: The Wright Fisher Model</strong><br />
September 24: The Wright-Fisher Model: Evolutionary Neutrality</p>
<p>September 26: The Wright-Fisher Mode 2: Types of Selection, Selective Sweeps</p>
<p><strong>Week Six: Multiple Loci - Measures of Genetic Variation</strong><br />
October 1: Nucleotide Diversity, the Site Frequency Spectrum</p>
<p>October 3: F statistics</p>
<p><strong>Week Seven: Biological Simulation and Population Size Changes</strong><br />
October 8: The SLiM Programming Language and <code>slimr</code></p>
<p>October 10: Population Bottlenecks, Population Expansion, and Genetic Diversity</p>
<p><strong>Week Eight:Biological Simulation and Population Size Changes Continued</strong><br />
October 15: Population Size Changes and the Site Frequency Spectrum</p>
<p>October 17: Fall Break</p>
<p><strong>Week Nine: Multiple Populations and Genetic Relatedness</strong><br />
October 22:Simulation with Multiple Populations - Migration</p>
<p>October 24:Simulation with Multiple Populations - Admixture, Local Adaptation</p>
<p><strong>Unit Three: Analyzing Genetic Data</strong><br />
How are population-scale genetic variants represented?<br />
How do we quantify relatedness between populations?<br />
Tests for selection, association testing, fine mapping.  
Phylogeny</p>
<p><strong>Week Ten: Association Testing</strong><br />
October 29: The Variant Call Format and population-scale data</p>
<p>October 31: GWAS, linkage disequilbrium, Fine Mapping</p>
<p><strong>Week Eleven: Population Structure and Phylogeny</strong><br />
November 5: Population structure: PCA, STRUCTURE, and clustering methods</p>
<p>November 7: The Coalescent - Inferring Timing of Selection</p>
<p><strong>Week Twelve: Constructing and Interpreting Phylogenies</strong><br />
November 12: Working with Phylogenetic Trees - the <code>ape</code> and <code>phytools</code> packages</p>
<p>November 14: Tree Comparison Methods, Advanced trees</p>
<p><strong>Week Thirteen: Independent/Small Group Projects</strong><br />
November 19: Work on Independent/Small Group Projects</p>
<p>November 21: Work on Independent/Small Group Projects</p>
<p><strong>Week Fourteen:</strong><br />
December 3: Semester Retrospective Discussion<br />
Evolutionary Methods in Other Fields (Linguistics, Economics, etc.)<br />
Current Directions in Population Genetics<br />
</p>
<p>December 5: Project Presentations and Discussion</p>
<p><strong>December 19</strong>: All revisions for weekly assignments due</p>

</div>
</div>
<hr>
<center> 
  <div class="footer">
      All illustrations <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY. </a>
      <br>
      All other materials <a href= "https://creativecommons.org/licenses/by/4.0/"> CC-BY </a> unless noted otherwise.
  </div>
</center>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="lecture-notes.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

</body>

</html>
